### ISSUE TITLE
Scaling of Weighting: Amount of Substance

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#217]
The weighting itself should be normalized to the typical weighting of the first species (former `NUM_EL_PER_PARTICLE`).

By that, we will operate with ratios-to-the-typical weighting instead of working with the weighting as an absolute number.
This should make reduces with real particle counters and formulas with basic constants like `Q_EL` way saver (for users that tend to calculate `pow( getCharge(1.0), 4);` ...).

**Comment by ax3l** [comment]
@Heikman that's the issue I am talking about.

The problem he described today is, that if one wants to use the `FieldTmp` in PIConGPU internal units, it is currently not a density but a mol density and only the `simUnit` of getUnit/openPMD makes it a SI density again.

I would propose we rename `particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE` to a `UNIT_AMOUNT_SUBSTANCE`. One could even give such scaled densities out via the [openPMD `unitDimension`](https://github.com/openPMD/openPMD-standard/blob/1.0.0/STANDARD.md#required-for-each-record), but for plugins we will probably just build a trait to get back to something reasonable.

**Comment by ax3l** [comment]
@n01r please take care to multiply your `FieldTmp` density with `particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE` for now and add a `/** \todo ... */` above the place where you use it, otherwise it is not really a "PIConGPU" scaled density you can work with.

**Comment by n01r** [comment]
Mh, I already noticed something like that.
I have to convert the density into g/cc anyway for my *lovely* TF formula.

**Comment by Heikman** [comment]
That sounds fine. So, to summarize, in the future we obtain the particle density by:
```
density = molDensity / UNIT_AMOUNT_SUBSTANCE
```
This is how we obatin the density internally *and* how we expose it to the file output.

**Comment by n01r** [comment]
Wait, what's the value of `UNIT_AMOUNT_SUBSTANCE`? The Avogadro constant or the inverse of it?   
(or sth completely different?)   
Because   
```
numberDensity = molarDensity * AVOGADRO_CONSTANT
```

**Comment by Heikman** [comment]
It is something completly different. Probably we'll choose to it to be `particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE`

**Comment by Heikman** [comment]
Your formula
```
numberDensity = molarDensity * AVOGADRO_CONSTANT
```
holds only if `molarDensity` and `numberDensity` are meant to be in SI units (as well as AVOGADRO_CONSTANT). 
In picongpu-units it is:
```
numberDensity = molarDensity / UNIT_AMOUNT_SUBSTANCE
```

**Comment by n01r** [comment]
huh ... unintuitive ^^;

**Comment by Heikman** [comment]
well in terms of units it's: `1 / m^3 = (mol / m^3)  /  mol`

**Comment by Heikman** [comment]
Wait, it's all wrong. This is right:
```
numberDensity = molarDensity * UNIT_AMOUNT_SUBSTANCE * AVOGADRO_CONSTANT
```

**Comment by n01r** [comment]
This looks a lot better now ^^ ... one more heretical question: is `AVOGADRO_CONSTANT` in units of `UNIT_AMOUNT_SUBSTANCE`? :cactus:

**Comment by Heikman** [comment]
No, in my formula it numerically equals to `6.022e23`. So it should better be called `AVOGADRO_CONSTANT_SI`.

**Comment by ax3l** [comment]
Exactly.
Just imagine that `UNIT_AMOUNT_SUBSTANCE` would convert a mol to something that is *numerically stable to work with* in PIConGPU units.

**Comment by n01r** [comment]
oh, alright, that clarifies it :)
